akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -233,6 +275,48 @@ object ShardCoordinator {
           emptyRebalanceResult
       } else emptyRebalanceResult
     }
+
+    private def possibleShardDestinations(currentShardAllocations: AllocationMap): AllocationMap =
+      if (!rollingUpdateInProgress) currentShardAllocations
+      else {
+        // rolling upgrade in progress, nodes with old version should be avoided
+        // note that this could return an empty map if the shard has not registered yet but a member has
+        // joined, in that case allocating on an old node is unavoidable
```

Comment:
@johanandren, can you clarify what you have in mind here? 

If I understand you correctly, such a scenario happens when a new node joins and register a new kind of entities. Something that didn't existed on the previous version. Is that what you mean?

Reply:
which makes me think that a new node may register a new kind of entities that the older nodes can't instantiate.

Reply:
No, it's not related to that, such a problem would be up to the user to not cause (if that is even possible to end up with).

The comment here is because in the allocation strategy we look at the Cluster().state which may not be in sync with the list of shards we got from the coordinator because of async + registering shards to coordinator could take some time after a node joined. 

So we could see a member with a version+1 in the cluster state, meaning we should not rebalance to old nodes, but then in the set of shard region we don't see the region for that member (yet), but we are also not allowed to return a "nowhere" in allocateShard, so in that scenario we must allow allocate to old member - even though we know rolling is in progress.

Reply:
Maybe we shouldn't care so much about this edge case and only consider a rolling update to be in place if one of the regions we're given has different versions.

The new node may never host this entity type due to DC/role.

Reply:
That's a good point chris, we should start from the regions when deciding this instead of looking at all members. I'll change that.